### ISSUE TITLE
IIS Handler Mappings error, Filename: \\?\C:\inetpub\wwwroot\snipe-it\public\web.config

### ISSUE THREAD

**Issue by wujoho** [snipe/snipe-it#6871]
#### Expected Behavior (or desired behavior if a feature request)

After adding site "Snipe IT" under IIS Sites, should be able to config Handler Mappings.

-----

#### Actual Behavior

When double click Handler Mappings receiving error:

_There was an error when performing this operation
Details: N/A
Filename: \\?\C:\inetpub\wwwroot\snipe-it\public\web.config
Error: N/A_

-----

#### Please confirm you have done the following before posting your bug report:

- [ YES ] I have enabled debug mode 
- [ YES ] I have read [checked the Common Issues page](https://snipe-it.readme.io/docs/common-issues)

-----
#### Provide answers to these questions:

- Is this a fresh install or an upgrade? 
fresh install

- Version of Snipe-IT you're running
downloaded via "git clone https://github.com/snipe/snipe-it your-folder" on 2019-03-30, should be 4.6.14 ?

- Version of PHP you're running
7.3.3

- Version of MySQL/MariaDB you're running
MySQL Server 5.7

- What OS and web server you're running Snipe-IT on
Windows 10 x64, IIS version 10.0.17763.1

- What method you used to install Snipe-IT (install.sh, manual installation, docker, etc)
followed step on https://snipe-it.readme.io/docs/downloading, via command line:
git clone https://github.com/snipe/snipe-it your-folder

- WITH DEBUG TURNED ON, if you're getting an error in your browser, include that error
HTTP Error 500.19 - Internal Server Error
The requested page cannot be accessed because the related configuration data for the page is invalid.
Detailed Error Information:
Module: IIS Web Core                     Requested URL: http://localhost:8080/
Error Code: 0x8007000d
Config File: \\?\C:\inetpub\wwwroot\snipe-it\public\web.config
More information: https://go.microsoft.com/fwlink/?LinkID=62293&IIS70Error=500,19,0x8007000d,17763

- What specific Snipe-IT page you're on, and what specific element you're interacting with to trigger the error
http://localhost:8080/

I put a test.php file in folder public, trying http://localhost:8080/test.php got the same HTTP Error 500.19 - Internal Server Error

The test.php file content is:
<?php
phpinfo();
?>

But if delete <rewrite>...</rewrite> section from public\web.config file, url http://localhost:8080/test.php works.

- If a stacktrace is provided in the error, include that too.
- Any errors that appear in your browser's error console.
- Confirm whether the error is reproducible on the demo: https://snipeitapp.com/demo.
- Include any additional information you can find in `storage/logs` and your webserver's logs.
- Include what you've done so far in the installation, and if you got any error messages along the way.
Have done steps 1 to 5 in page: https://snipe-it.readme.io/docs/installation

- Indicate whether or not you've manually edited any data directly in the database
No

Please do not post an issue without answering the related questions above. If you have opened a different issue and already answered these questions, answer them again, once for every ticket. It will be next to impossible for us to help you.

https://snipe-it.readme.io/docs/getting-help

**Comment by BrettFagerlund** [comment]
Please confirm that you have URL Rewrite installed

**Comment by wujoho** [comment]
Yes, Visual C++ 2010, 2012, 2013, 2017 x86 and x64 are installed, and I can open Handler Mappings for Default Web Site under IIS

**Comment by HinchK** [comment]
There are a few gotchas with regard to IIS, let's work through them:

Check Permissions,
https://snipe-it.readme.io/docs/windowsiis#section-fix-permissions
_note: IUSR is a LOCAL user.  Domains will show additional IUSR_XXXX users - don't use these.  IIRC, the snipe-it backup too also requires the owner of "C:\Windows\Temp" to be  this IUSR as well.   (*Owner and All-Privs)_

Handler Mappings:
https://snipe-it.readme.io/docs/windowsiis#section-set-additional-handler-mappings


IIS is the toughest to get off the ground when a roadblock presents itself. 
 * hm, didn't work, let's follow the logs [+20MIN, Event Viewer Fiddling]
 * backups don't work (keep smashing pathing issues to the mysqldump.exe.....only to find out that "C:\WINDOWS\Temp" also needs IUSR ownership and full-perms.
 * how do I schedule my notification checks using Scheduled Tasks instead of cronjobs.?

Question: Scheduled Tasks can't run these "php artisan" CLI commands work when they fire?   How is this supposed to work?   
Answer: You have to write .BAT batch files  containing the php commands you wish to run.   There's no other way around this.



For this reason, this documentation here: https://snipe-it.readme.io/docs/windowsiis is something that we would like to be as bulletproof as possible.   If you see anything in this documentation that does NOT work for you @wujoho, or if we missed a step that was critical to your installation, please let us know here, or make a "suggested edit" request to the documents themselves.

Best of luck!

**Comment by wujoho** [comment]
> 
> 
> Please confirm that you have URL Rewrite installed

@BrettFagerlund  sorry, I though URL Rewrite is the same component as Visual C++ Redistributable, they are not.

Today, I installed URL Rewrite and the old issue with Handler Mappings and web.config fixed, thanks.

@HinchK I have done the permission settings, the same as the ones mentioned in https://snipe-it.readme.io/docs/windowsiis#section-fix-permissions
Now when I run http://localhost:8080 the site stop at url http://localhost:8080/setup and page shows error, I checked folder storage has been granted modify permission:

(1/1)                                     UnexpectedValueException                                                                  The  stream or file "C:\inetpub\wwwroot\snipe-it\storage/logs/laravel.log"  could not be opened: failed to open stream: Permission denied
--
in StreamHandler.php line 107
at StreamHandler->write(array('message'  => 'UnexpectedValueException: The stream or file  "C:\\inetpub\\wwwroot\\snipe-it\\storage/logs/laravel.log" could not be  opened: failed to open stream: Permission denied in  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\StreamHandler.php:107Stack  trace:#0  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\AbstractProcessingHandler.php(39):  Monolog\\Handler\\StreamHandler->write(Array)#1  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Logger.php(344):  Monolog\\Handler\\AbstractProcessingHandler->handle(Array)#2  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Logger.php(707):  Monolog\\Logger->addRecord(400, Object(UnexpectedValueException),  Array)#3  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Log\\Writer.php(203):  Monolog\\Logger->error(Object(UnexpectedValueException), Array)#4  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Log\\Writer.php(114):  Illuminate\\Log\\Writer->writeLog(\'error\',  Object(UnexpectedValueException), Array)#5  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Support\\Facades\\Facade.php(221):  Illuminate\\Log\\Writer->error(Object(UnexpectedValueException))#6  C:\\inetpub\\wwwroot\\snipe-it\\app\\Exceptions\\Handler.php(39):  Illuminate\\Support\\Facades\\Facade::__callStatic(\'error\', Array)#7  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Http\\Kernel.php(314):  App\\Exceptions\\Handler->report(Object(UnexpectedValueException))#8   C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Http\\Kernel.php(118):   Illuminate\\Foundation\\Http\\Kernel->reportException(Object(UnexpectedValueException))#9  C:\\inetpub\\wwwroot\\snipe-it\\public\\index.php(58):  Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))#10  {main}', 'context' => array(), 'level' => 400, 'level_name' => 'ERROR', 'channel' => 'production', 'datetime' => object(DateTime), 'extra' => array(),  'formatted' => '[2019-04-06 12:29:01] production.ERROR:  UnexpectedValueException: The stream or file  "C:\\inetpub\\wwwroot\\snipe-it\\storage/logs/laravel.log" could not be  opened: failed to open stream: Permission denied in  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\StreamHandler.php:107Stack  trace:#0  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\AbstractProcessingHandler.php(39):  Monolog\\Handler\\StreamHandler->write(Array)#1  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Logger.php(344):  Monolog\\Handler\\AbstractProcessingHandler->handle(Array)#2  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Logger.php(707):  Monolog\\Logger->addRecord(400, Object(UnexpectedValueException),  Array)#3  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Log\\Writer.php(203):  Monolog\\Logger->error(Object(UnexpectedValueException), Array)#4  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Log\\Writer.php(114):  Illuminate\\Log\\Writer->writeLog(\'error\',  Object(UnexpectedValueException), Array)#5  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Support\\Facades\\Facade.php(221):  Illuminate\\Log\\Writer->error(Object(UnexpectedValueException))#6  C:\\inetpub\\wwwroot\\snipe-it\\app\\Exceptions\\Handler.php(39):  Illuminate\\Support\\Facades\\Facade::__callStatic(\'error\', Array)#7  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Http\\Kernel.php(314):  App\\Exceptions\\Handler->report(Object(UnexpectedValueException))#8   C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Http\\Kernel.php(118):   Illuminate\\Foundation\\Http\\Kernel->reportException(Object(UnexpectedValueException))#9  C:\\inetpub\\wwwroot\\snipe-it\\public\\index.php(58):  Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))#10  {main}  '))in AbstractProcessingHandler.php line 39
at AbstractProcessingHandler->handle(array('message'  => 'UnexpectedValueException: The stream or file  "C:\\inetpub\\wwwroot\\snipe-it\\storage/logs/laravel.log" could not be  opened: failed to open stream: Permission denied in  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\StreamHandler.php:107Stack  trace:#0  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\AbstractProcessingHandler.php(39):  Monolog\\Handler\\StreamHandler->write(Array)#1  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Logger.php(344):  Monolog\\Handler\\AbstractProcessingHandler->handle(Array)#2  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Logger.php(707):  Monolog\\Logger->addRecord(400, Object(UnexpectedValueException),  Array)#3  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Log\\Writer.php(203):  Monolog\\Logger->error(Object(UnexpectedValueException), Array)#4  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Log\\Writer.php(114):  Illuminate\\Log\\Writer->writeLog(\'error\',  Object(UnexpectedValueException), Array)#5  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Support\\Facades\\Facade.php(221):  Illuminate\\Log\\Writer->error(Object(UnexpectedValueException))#6  C:\\inetpub\\wwwroot\\snipe-it\\app\\Exceptions\\Handler.php(39):  Illuminate\\Support\\Facades\\Facade::__callStatic(\'error\', Array)#7  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Http\\Kernel.php(314):  App\\Exceptions\\Handler->report(Object(UnexpectedValueException))#8   C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Http\\Kernel.php(118):   Illuminate\\Foundation\\Http\\Kernel->reportException(Object(UnexpectedValueException))#9  C:\\inetpub\\wwwroot\\snipe-it\\public\\index.php(58):  Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))#10  {main}', 'context' => array(), 'level' => 400, 'level_name' => 'ERROR', 'channel' => 'production', 'datetime' => object(DateTime), 'extra' => array(),  'formatted' => '[2019-04-06 12:29:01] production.ERROR:  UnexpectedValueException: The stream or file  "C:\\inetpub\\wwwroot\\snipe-it\\storage/logs/laravel.log" could not be  opened: failed to open stream: Permission denied in  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\StreamHandler.php:107Stack  trace:#0  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Handler\\AbstractProcessingHandler.php(39):  Monolog\\Handler\\StreamHandler->write(Array)#1  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Logger.php(344):  Monolog\\Handler\\AbstractProcessingHandler->handle(Array)#2  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\monolog\\monolog\\src\\Monolog\\Logger.php(707):  Monolog\\Logger->addRecord(400, Object(UnexpectedValueException),  Array)#3  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Log\\Writer.php(203):  Monolog\\Logger->error(Object(UnexpectedValueException), Array)#4  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Log\\Writer.php(114):  Illuminate\\Log\\Writer->writeLog(\'error\',  Object(UnexpectedValueException), Array)#5  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Support\\Facades\\Facade.php(221):  Illuminate\\Log\\Writer->error(Object(UnexpectedValueException))#6  C:\\inetpub\\wwwroot\\snipe-it\\app\\Exceptions\\Handler.php(39):  Illuminate\\Support\\Facades\\Facade::__callStatic(\'error\', Array)#7  C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Http\\Kernel.php(314):  App\\Exceptions\\Handler->report(Object(UnexpectedValueException))#8   C:\\inetpub\\wwwroot\\snipe-it\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Http\\Kernel.php(118):   Illuminate\\Foundation\\Http\\Kernel->reportException(Object(UnexpectedValueException))#9  C:\\inetpub\\wwwroot\\snipe-it\\public\\index.php(58):  Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))#10  {main}  '))in Logger.php line 344
at Logger->addRecord(400, object(UnexpectedValueException), array())in Logger.php line 707
at Logger->error(object(UnexpectedValueException), array())in Writer.php line 203
at Writer->writeLog('error', object(UnexpectedValueException), array())in Writer.php line 114
at Writer->error(object(UnexpectedValueException))in Facade.php line 221
at Facade::__callStatic('error', array(object(UnexpectedValueException)))in Handler.php line 39
at Handler->report(object(UnexpectedValueException))in HandleExceptions.php line 81
at HandleExceptions->handleException(object(UnexpectedValueException))

**Comment by wujoho** [comment]
@HinchK For folder _storage_, I changed IUSR permission from _Modify_ to _Full control_ still having permission issue, then I added a new user _Everyone_ with permission _Modify_, the site works.

Then I test the Snipe-IT Pre-Flight page and it goes to http://localhost:8080/setup without any error, all Pre-Flight Check points are good.
I click button _Next: Create Database Tables_, but after a while, the page jumps to an error page http://localhost:8080/setup/migrate, showing: 
**Whoops, looks like something went wrong.**

Not sure how to fix this, please help @BrettFagerlund

**Comment by wujoho** [comment]
I go to page http://localhost:8080/setup and click button _Next: Create Database Tables_ again, this time there is no error but showing a warning message at page http://localhost:8080/setup/migrate:
_Create Database Tables
There was nothing to migrate. Your database tables were already set up!

Migration output:

Nothing to migrate._

Then I continue click button _Next: Create User_, _Next: Create User_, all good.

So basically, there were two issues:
1. _Modify_ permission for _IUSR_ at folder _storage_ is not enough, I have to add _Everyone_ with _Modify_ permission
2. At Pre-Flight page http://localhost:8080/setup click button _Next: Create Database Tables_ leads to error page http://localhost:8080/setup/migrate, have to go back http://localhost:8080/setup and click button _Next: Create Database Tables_ again to fix.

Thanks

**Comment by okhawa** [comment]
> I go to page http://localhost:8080/setup and click button _Next: Create Database Tables_ again, this time there is no error but showing a warning message at page http://localhost:8080/setup/migrate:
> _Create Database Tables
> There was nothing to migrate. Your database tables were already set up!
> 
> Migration output:
> 
> Nothing to migrate._
> 
> Then I continue click button _Next: Create User_, _Next: Save User_, all good.
> 
> So basically, there were two issues:
> 
> 1. _Modify_ permission for _IUSR_ at folder _storage_ is not enough, I have to add _Everyone_ with _Modify_ permission
> 2. At Pre-Flight page http://localhost:8080/setup click button _Next: Create Database Tables_ leads to error page http://localhost:8080/setup/migrate, have to go back http://localhost:8080/setup and click button _Next: Create Database Tables_ again to fix.
> 
> Thanks

I need to add "Everyone" with Modify permission on Storage folder also, IUSR with Modify permission not enough.

**Comment by BrettFagerlund** [comment]
You could use wireshark to determine the actual user that was trying to write, but with Win10/FastCGI etc. It's highly likely it's running under the application identity rather than IUSR. I would swap that everyone permission across to the app pool user, or the IIS_IUSRS group depending on your knowledge of IIS .
